service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java

```diff
@@ -63,13 +73,47 @@ public RunSetsApiController(
 
   @Override
   public ResponseEntity<RunSetStateResponse> postRunSet(RunSetRequest request) {
-    if (!requestIsValid(request, this.cbasApiConfiguration)) {
-      return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+    Gson gson = new Gson();
+
+    // request validation
+    Optional<ResponseEntity<RunSetStateResponse>> validateRequestResponse =
+        validateRequest(request, this.cbasApiConfiguration.getRunSetsMaximumRecordIds());
+    if (validateRequestResponse.isPresent()) {
+      return validateRequestResponse.get();
     }
 
-    // Store the method
-    UUID methodId = UUID.randomUUID();
+    // Fetch WDS Records and keep track of errors while retrieving records
+    String recordType = request.getWdsRecords().getRecordType();
+    List<String> recordIds = request.getWdsRecords().getRecordIds();
+    recordRecordsInRequest(recordIds.size());
 
+    ArrayList<RecordResponse> recordResponses = new ArrayList<>();
+    HashMap<String, String> recordIdsWithError = new HashMap<>();
+    for (String recordId : recordIds) {
+      try {
+        recordResponses.add(wdsService.getRecord(recordType, recordId));
+      } catch (ApiException e) {
+        log.warn("Record lookup for Record ID {} failed.", recordId, e);
+        ErrorResponse error = gson.fromJson(e.getResponseBody(), ErrorResponse.class);
+        String errorMsg;
+        if (error == null) {
+          errorMsg = e.getMessage();
+        } else {
+          errorMsg = error.getMessage();
+        }
```

Comment:
Perhaps this could be extracted into its own "find useful error message" utility function?

Reply:
Yeah good point. That way adding a try catch inside method would be cleaner than adding it here.